### PR TITLE
Add support for focusing on named goals using brackets.

### PR DESCRIPTION
**Pull request by Zimmi48**
**Kind:** feature

Grants the first part of #8026

- [x] Added / updated test-suite

- [x] Corresponding documentation was added / updated
  - [x] Including any warning and error messages added / removed / modified).
- [x] Entry added in CHANGES.

### CODE REVIEW

File: ide/coq_lex.mll
Comment (Zimmi48) [inline ide/coq_lex.mll]: For now, this is a clear under-approximation. I'm not against some help in making it less so.

**Review by ppedrot — APPROVED** [review]
LGTM. Reviewing this patch made me realize that we're still using this horrible Obj hack in the Proof module. I'll submit a fix for it.

**Comment by Zimmi48** [comment]
@ppedrot What do you think of the parsing under-approximation in CoqIDE? Ideally, we should stop requesting editors to parse Coq files (cf. SerAPI which does the splitting of sentences on its own), but in the meantime what to do? Keep it an under-approximation or improve this? Do you have suggestions?

**Comment by ppedrot** [comment]
The current splitting algorithm is already an approximation IIRC, so I don't think it really hurts to extend it a bit. We could probably defer it to coq itself, but then I am slightly concerned about the potential cost to it, as it would introduce latency if done improperly. Anyways we can always enhance this after this PR.

**Review by cpitclaudel — APPROVED** [review]
Docs look good.

**Comment by cpitclaudel** [comment]
Do recent Coq versions have a solution for this in Ltac?  That is, can you run a piece of Ltac in the context of a named evar? The one solution I'm familiar with is `instantiate (1 := ltac:(…))`, which is a bit unwieldy.

**Comment by cpitclaudel** [comment]
> but you cannot use this as part of a larger Ltac macro (goal selectors are not technically part of Ltac).

Right, that's what I meant.  Could your code be extended to allow `instantiate ([x] := ltac:(…))`

**Comment by Zimmi48** [comment]
Sorry but my previous answer was quite incomplete, you can do the following:

```coq
Goal exists n : nat, n = n.
eexists ?[x]; only [x]: exact 0.
```

Basically you add the `only` keyword to be able to use a goal selector inside Ltac (documented here https://coq.inria.fr/refman/proof-engine/ltac.html#coq:tacv.only and available since 8.6). Sorry for forgetting about this earlier on. I would definitely recommend using this over `instantiate` whenever you can.

**Comment by herbelin** [comment]
@cpitclaudel, there is already syntax for that:
```coq
Goal exists x, x = 1.
eexists ?[x].
instantiate (x := ltac:(apply S)).
```
Note ongoing discussions on the status of `instantiate (n:=...)`. In particular, `instantiate (x:=...)` should not be the same as `instantiate (n:=...)`. The first one is in the context of the evar. The second one should morally be in the context of the goal in which the evar appears (since the evar is then in general subject to a substitution).

**Comment by cpitclaudel** [comment]
Thanks @Zimmi48 and @herbelin , this is great to know :)

**Comment by Zimmi48** [comment]
@ppedrot: @mattam82 is only a code owner because of the trivial change in `vernacentries.ml` so I don't think you need to wait for his review before merging.

**Comment by ppedrot** [comment]
All right, let's do it then.

**Pull request merged**
